### PR TITLE
APPS-1225-add-file-object-metadata

### PR DESCRIPTION
**Pull request by YuxinShi0423**
This PR is to add `metadata` field in the CWL File Object. It is not explicitly supported by CWL v1.2.0 standard, however, cwltool could parse it as a valid map, so users could couple some metadata along with the supplied File object, and propagate them to the Javascript expressions.

Full support of File metadata is expected later this year in CWL v1.2.1. This PR is a quick implementation of metadata that is used only on DNAnexus platform. 

The metadata field is stringified during compilation and then deserialized back to the JSON map in the execution env on the DNAnexus platform. 

The purpose of stringifying the metadata field is that:

- when translating each field its data type is fixed. For example, the File `location` field is expected to be a URI and the `size` is required to be an integer. But the metadata field will have no such limit, in other words, any JSON type might appear.

- Therefore, stringifying the whole metadata field can help safely pack all the metadata values in their original format, so later cwltool can properly process it when executing the source code.

### CODE REVIEW

**Review by Gvaihir — COMMENTED** [review]
Partial review.
Please add tests and update RELEASE_NOTES

File: src/main/scala/dx/cwl/CwlValue.scala [hunk 1]
```diff
@@ -915,7 +920,8 @@ object FileValue {
         },
         translateOptionalArray(file.getSecondaryFiles).map(PathValue.apply),
         translateOptional(file.getFormat),
-        translateOptional(file.getContents)
+        translateOptional(file.getContents),
+        None
```
Comment (Gvaihir) [inline src/main/scala/dx/cwl/CwlValue.scala:932]: Why here defaults to `None`? Was there an implementation of smth like `FileImpl.getMetadata()` in the new `cwlJava`?
What's `FileImpl` - just a file uri on the disk?

File: src/main/scala/dx/cwl/CwlValue.scala [hunk 2]
```diff
@@ -975,7 +982,8 @@ object FileValue {
                 throw new Exception(s"invalid secondaryFiles value ${other}")
             },
             fields.get("format").map(PathValue.unwrapString),
-            fields.get("contents").map(PathValue.unwrapString)
+            fields.get("contents").map(PathValue.unwrapString),
+            fields.get("metadata").map(_.prettyPrint)
```
Comment (Gvaihir) [inline src/main/scala/dx/cwl/CwlValue.scala]: What if `None` returned from `get` here?

**Review by Gvaihir — CHANGES_REQUESTED** [review]
Minor changes, otherwise good. Pls don't forget unit tests - I know changes might be very surgical, but tests are good even for the sake of regression down the road

File: src/main/scala/dx/cwl/CwlValue.scala (hunk 1 above)
Comment (YuxinShi0423) [reply src/main/scala/dx/cwl/CwlValue.scala:932]: `FileImpl` is part of cwljava to parse the source workflow file and generate `File` object. The current cwl standard does not have metadata as an attr of File, but it is requested by BMS. See my comment in the PR description and https://jira.internal.dnanexus.com/browse/APPS-1225

File: src/main/scala/dx/cwl/CwlValue.scala (hunk 2 above)
Comment (YuxinShi0423) [reply src/main/scala/dx/cwl/CwlValue.scala]: `fields.get("metadata").map(_.prettyPrint)` will return `None` if `metadata` field does not exist.

Comment (Gvaihir) [reply src/main/scala/dx/cwl/CwlValue.scala]: I mean I get _what_ is returned. The question is if `_` is `None` then `None.prettyPrint` should throw an exception at runtime. As I mentioned - cover this corner case with unit tests and check

File: src/test/scala/dx/cwl/CwlValueTest.scala [hunk 3]
```diff
@@ -27,6 +27,29 @@ class CwlValueTest extends AnyFlatSpec with Matchers {
     v shouldBe IntValue(5)
   }
 
+  it should "deserialize the input JSON to File " in {
```
Comment (Gvaihir) [inline src/test/scala/dx/cwl/CwlValueTest.scala]: You introduced metadata into implementation of several `apply` methods of `FileValue`, but testing only one here. For example `apply(file: FileImpl)` where you assign `None` to metadata. But `deserialize` method has `get` on metadata which might be `None`. Try to cover that corner case to make sure you don't get `None.prettyPrint` at some point

File: src/test/scala/dx/cwl/EvaluatorTest.scala [hunk 4]
```diff
@@ -89,7 +105,7 @@ class EvaluatorTest extends AnyWordSpec with Matchers {
                   streamable = false,
                   loadContents = true,
                   loadListing = LoadListing.No
-              ) -> FileValue(location = Some(f.toString)),
+              ) -> FileValue(location = Some(f.toString), metadata = Some(fileMeta.prettyPrint)),
```
Comment (Gvaihir) [inline src/test/scala/dx/cwl/EvaluatorTest.scala]: As before - cover one case when metadata is `None`. Here or in `CwlValueTest`

**Review by Gvaihir — CHANGES_REQUESTED** [review]

**Comment by Gvaihir** [comment]
`RELEASE_NOTES` pending

File: src/main/scala/dx/cwl/CwlValue.scala (hunk 2 above)
Comment (YuxinShi0423) [reply src/main/scala/dx/cwl/CwlValue.scala]: > Expression type:
> 
> Option[String]
> Symbol signature:
> 
> final def map[B](f: JsValue => B): Option[B]
> Returns a [scala.Some](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) containing the result of applying f to this [scala.Option](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)'s value if this [scala.Option](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html) is nonempty. Otherwise return None.
> 
> This is equivalent to:
> 
> option match {
>   case Some(x) => Some(f(x))
>   case None    => None
> }

Comment (YuxinShi0423) [reply src/main/scala/dx/cwl/CwlValue.scala]: But you mean it would still do `None.prettyprint`? @Gvaihir

**Comment by YuxinShi0423** [comment]
> `RELEASE_NOTES` pending

Shall we make this `metadata` visible to the users at this stage? @commandlinegirl @emiloslavsky

**Comment by emiloslavsky** [comment]
@YuxinShi0423 the introductory comment needs to be updated; it currently says "The metadata field is stringified during compilation and then deserialized back to the JSON map in the execution env on the DNAnexus platform...."

Please document metadata support as "under development, pending inclusion in CWL standard".

File: src/main/scala/dx/cwl/CwlValue.scala (hunk 2 above)
Comment (Gvaihir) [reply src/main/scala/dx/cwl/CwlValue.scala]: I thought so.

**Review by Gvaihir — APPROVED** [review]
